### ISSUE TITLE
When can't MatchRoutes be used for misses?

### ISSUE THREAD

**Issue by pshrmn** [ReactTraining/react-router#4136]
Basically, I'm fairly convinced that `<Miss>` doesn't need to exist and want to figure out a situation where a `<Miss>`'s functionality can't be replaced with a `<MatchGroup>` in order to prove otherwise.

```js
const Page = () => (
  <div>
    <div className="before">
      <h2>Before Components</h2>
      <Match pattern='before' component={Before} />
      <Match pattern='both' component={BeforeBoth} />
    </div>
    <div className="other-stuff">
      <Link to="/foo">Foo</Link>
    </div>
    <div className="after">
      <h2>After Components</h2>
      <Match pattern='after' component={After} />
      <Match pattern='both' component={AfterBoth} />
    </div>
    <Miss />
  </div>
)
```

That could be transformed to:

```js
const Page = () => (
  <MatchGroup miss={NoMatch}>
    <Match pattern='before' component={Before} />
    <Match pattern='after' component={After} />
    <Match pattern='both' component={Both} />
  </MatchGroup>
)
```

And the original page structure would be rendered with a container component.

```js
const PageContainer = (props) => 
  <div>
    <div className="before">
      <h2>Before Components</h2>
      {props.before}
    </div>
    <div className="other-stuff">
      <Link to="foo">Foo</Link>
    </div>
    <div className="after">
      <h2>After Components</h2>
      {props.after}
    </div> 
    {props.miss}
  </div>
)

const Before = () => (
  <PageContainer
    before={<p>Before</p>} />
)

const After = () => (
  <PageContainer
    after={<p>After</p>} />
)

const Both = () => (
  <PageContainer 
    before={<p>Before</p>}
    after={<p>After</p>} />
)

const NoMatch = () => (
  <PageContainer
    miss={<p>Miss!</p>} />
)
```
Where that would run into issues is if there were multiple `<Match>`es that match a given location. The `exactly` prop should help with that.

**Comment by alisd23** [comment]
Interesting idea. I suppose it's a similar idea to the discussion in #3914, except this is specifically concerned with `MatchGroup`.

But what issues would this solve? First-pass miss rendering?

**Comment by pshrmn** [comment]
Technically the existence of `<MatchGroup>` by itself should allow for first pass rendering as long as you don't use `<Miss>` components outside of it. It doesn't right now, but you could manipulate the context to do so. That said, with a miss prop instead of a component, the first pass miss rendering is simpler.

If the `<Miss>` is eliminated it would simplify the codebase. Without a `<Miss>`, `<Match>`es don't have to register themselves, so the `<RegisterMatch>` and `<MatchProvider>` components would no longer be necessary. The `<Match>` would just need to expose a context variable of its `match` so that child `<Match>`es and `<MatchGroup>`s can correctly match relative to its `pathname`.

I could very well be overlooking a situation that cannot be rewritten without a `<Miss>`, but nothing has come to mind yet.

**Comment by Ciantic** [comment]
How would you replace the misses in this example with MatchGroup miss? Or can I nest match groups to achieve same thing?

```
const OtherApp = ({pathname}) => (
  <OtherUI>
      <Match pattern={`${pathname}/about`} component={OtherAbout} />
      <Miss component={OtherMiss} />
  </OtherUI>
)

const FancyApp = ({pathname}) => (
  <FancyUI>
    <Match pattern={`${pathname}/about`} component={FancyAbout} />
    <Miss component={FancyMiss} />
  </FancyUI>
)

const ThirdApp = ({pathname}) => (
  <ThirdUI>
    <Match pattern={`${pathname}/about`} component={ThirdAbout} />
    <Miss component={ThirdMiss} />
  </ThirdUI>
)

const App = () => (
  <MatchGroup>
    <Match pattern='/other' component={OtherApp} />
    <Match pattern='/fancy' component={FancyApp} />
    <Match pattern='/third' component={ThirdApp} />
  </MatchGroup>
)
```

**Comment by pshrmn** [comment]
@Ciantic You would just use nested match groups. The only time that match groups become complicated is when the match and miss components wouldn't share the same parent, and container components should mitigate this.

**Comment by ryanflorence** [comment]
`MatchGroup` has been a backup plan in my mind for months for `Miss`.

The problem with MatchGroup though is composability:

```js
const MatchUsers = (props) => <Match pattern="/users" {...props}/>

// won't work :(
<MatchGroup>
  <MatchUsers/>
</MatchGroup>
```

Dynamic compostability breaks down because `MatchGroup` moves us one step away from dynamic routing to static routing. Each child of `MatchGroup` necessarily must have a `pattern` prop.

That one caveat may seem insignificant ("just require that constraint, should be fine!") but then we've got one foot over the line to static route config, which will keep creeping over until we end up farther away from the simplicity and non-surprising behavior of "just components".

So, I'm kind of thinking of this instead of `MatchGroup` have:

```
<Match patterns={[]}>{(matched, matchedPattern, props) => {}}</Match>
```

**Comment by pshrmn** [comment]
I'm guilty of liking a little bit of route config.

Is the `patterns` array something like the example [here](https://github.com/ReactTraining/react-router/issues/3914#issue-177866764)? I'm having trouble imagining anything else.

Also, if this change is made will the `<Miss>` component still exist?

**Comment by jedwards1211** [comment]
@ryanflorence I'm inclined toward `<Match patterns={[]}>` as well although I think we need to have an additional `exactPatterns` parameter to handle matching `/` for example.

Also, maybe it would be nice to be able to tag it with `consume={false}` so that we can do the following:

```es6
const InAppShell = (): React.Element<any> => (
  <AppShell>
    <Match exactly pattern="/" component={Home} />
    <Match pattern="/about" component={About} />
    <Match pattern="/contact" component={Contact} />
  </AppShell>
)

const App = (): React.Element<any> => (
  <div>
    <Match consume={false} exactlyPatterns={["/"]} patterns={["/about", "/contact"]} component={InAppShell} />
    <Miss component={NotFound} />
  </div>
)
```

@pshrmn static route config has its advantages but it's not actually impossible to implement via these new dynamic APIs.  And I think dynamic routing is the way forward because I've run into situations before where I needed to look at my redux state to determine which child routes to render, and with v2 that was fairly awkward.

**Comment by jedwards1211** [comment]
@ryanflorence does `<Miss>` render even if it's inside a parent `<Match>` that matched?  (i.e. if no `<Match>`es directly under its parent `<Match>` matched?)

I'm starting to worry that between v2 and v4 there's really no ideal solution to all problems...

**Comment by jedwards1211** [comment]
@ryanflorence @pshrmn an alternative to `<Miss>` that only requires one pass and still has the simplicity of "just components" would be negative matching:

```es6
const App = (): React.Element<any> => (
  <div>
    <Match pattern="/about" component={About} />
    <Match pattern="/contact" component={Contact} />
    <Match notPatterns={["/about", "/contact"]} component={FourOhFour} />
  </div>
)
```

It's kind of unpleasant to have to put all possible patterns in that array, but I think these kind of tradeoffs are probably unavoidable if we choose to go the route of "just components".

**Comment by jedwards1211** [comment]
Okay I made my own variation of `Match` that does exactly what I want.  I put `register={false}` on the top-level match for the app shell so that it doesn't put a match on context for its component, and the sub-Matches work.

What do you think?

```es6
function getPatterns(pattern: ?string, patterns: ?Array<string>): Array<string> {
  const result = []
  if (pattern) result.push(pattern)
  if (patterns) result.push(...patterns)
  return result
}

class Match extends React.Component {
  static defaultProps = {
    exactly: false,
    register: true,
  }

  static contextTypes = {
    match: PropTypes.object
  }

  getMatch(location) {
    const patterns = getPatterns(this.props.pattern, this.props.patterns)
    const notPatterns = getPatterns(this.props.notPattern, this.props.notPatterns)
    const exactPatterns = getPatterns(this.props.exactPattern, this.props.exactPatterns)
    const notExactPatterns = getPatterns(this.props.notExactPattern, this.props.notExactPatterns)
    const { match:matchContext } = this.context
    const parent = matchContext && matchContext.parent

    for (let notExactPattern of notExactPatterns) {
      if (matchPattern(notExactPattern, location, true, parent)) return null
    }
    for (let notPattern of notPatterns) {
      if (matchPattern(notPattern, location, false, parent)) return null
    }
    for (let exactPattern of exactPatterns) {
      const match = matchPattern(exactPattern, location, true, parent)
      if (match) return match
    }
    for (let pattern of patterns) {
      const match = matchPattern(pattern, location, false, parent)
      if (match) return match
    }
    if (!patterns.length && !exactPatterns.length) return {params: {}, exact: false, pathname: ""}
  }

  render() {
    return (
      <LocationSubscriber>
        {(location) => {
          const {
            children,
            render,
            component:Component,
            register,
          } = this.props

          const match = this.getMatch(location)
          const props = { ...match, location }
          const content = children
            ? children({matched: Boolean(match), ...props})
            : match
              ? render
                ? render(props)
                : <Component {...props} />
              : null

          if (register) {
            return (
              <RegisterMatch match={match}>
                <MatchProvider match={match}>
                  {content}
                </MatchProvider>
              </RegisterMatch>
            )
          }
          return content
        }}
      </LocationSubscriber>
    )
  }
}
```

### Example Usage
```es6
const InAppShell = (): React.Element<any> => (
  <AppShell>
    <Match exactPattern="/" component={Home} />
    <Match pattern="/about" component={About} />
    <Match pattern="/contact" component={Contact} />
  </AppShell>
)

const App = (): React.Element<any> => (
  <div>
    <Match exactPattern="/" patterns={["/about", "/contact"]} register={false} component={InAppShell} />
    <Match notExactPattern="/" notPatterns={["/about", "/contact"]} component={NotFound} />
  </div>
)
```

**Comment by pshrmn** [comment]
Having seen `<MatchRoutes>`, I don't think that I will use anything else. Realistically, I think that the `<Match>` and `<Miss>` components could be dropped (and rename `<MatchRoutes>` to `<Match>`).  The only big change this would cause is that you compose the matched component instead of the `<Match>` component when you want to do thinks like restrict pages to authorized users.

This would force patterns that are concerned with the same part of the UI to be grouped, while allowing for `<Match>`es with the same parent to not affect one another.

Basically, I like what @acdlite mentioned in #3914. The one thing that I haven't been able to wrap my head around is the tabbed settings example mentioned by @timdorr. I'd have to see some example markup but I feel like container components would be able to be used for that.

@jedwards1211 I think that you're over-complicating this. Just use the `missComponent` prop of `<MatchRoutes>`.

**Comment by jedwards1211** [comment]
@pshrmn perhaps so, but there's no reason we all have to use the same thing.  I suggested negative matching because it would be more flexible than `MatchRoutes` in certain edge cases; you would be able to push some `Match`es downward in the component hierarchy without breaking the miss handling.

**Comment by Lube** [comment]
MatchRoutes without dinamic routing is such a bad compromise, is there anyway for us to chime in on the progress of the v4 api @ryanflorence

**Comment by pshrmn** [comment]
With the `<Miss>`-less `v4-alt` branch having been merged into `v4` and a beta [releasing soon](https://twitter.com/mjackson/status/817811006982238209), I think that it is safe to close this issue.
